logstash-core/lib/logstash/compiler/lscl.rb

```diff
@@ -0,0 +1,555 @@
+      if defined?(super)
+        return super
+      end
+
+      exprs = self.recursive_select(Condition, Selector).map(&:expr)
+
+      raise "Exprs should only have one part!" if exprs.size != 1
+      exprs.first
+    end
+  end
+
+  module NegativeExpression
+    include Helpers
+
+    def expr
+      exprs = self.recursive_select(Condition, Selector).map(&:expr)
+      raise "Negative exprs should only have one part!" if exprs.size != 1
+      jdsl.eNot(source_meta, exprs.first)
+    end
+  end
+
+  module ComparisonExpression
+    include Helpers
+
+    def expr
+      lval, comparison_method, rval = self.recursive_select(Selector, Expression, ComparisonOperator, Number, String).map(&:expr)
+      comparison_method.call(source_meta, lval, rval)
+    end
+  end
+
+  module InExpression
+    include Helpers
+
+    def expr
+      item, list = recursive_select(RValue)
+      jdsl.eIn(source_meta, item.expr, list.expr)
+    end
+  end
+
+  module NotInExpression
+    include Helpers
+
+    def expr
+      item, list = recursive_select(RValue)
+      jdsl.eNot(source_meta, jdsl.eIn(item.expr, list.expr))
+    end
+  end
+
+  # Not implemented because no one uses this
+  class MethodCall < Node; end
+
+  class RegexpExpression < Node
+    def expr
+      selector, operator_method, regexp = recursive_select(Selector, LogStash::Compiler::LSCL::AST::RegExpOperator, LogStash::Compiler::LSCL::AST::RegExp).map(&:expr)
+
+      raise "Expected a selector #{text_value}!" unless selector
+      raise "Expected a regexp #{text_value}!" unless regexp
+
+      operator_method.call(source_meta, selector, regexp);
+    end
+  end
+
+  module BranchOrPlugin; end
+
+  module ComparisonOperator
+    include Helpers
+
+    def expr
+      case self.text_value
+      when "=="
+        jdsl.method(:eEq)
+      when "!="
+        jdsl.method(:eNeq)
+      when ">"
+        jdsl.method(:eGt)
+      when "<"
+        jdsl.method(:eLt)
+      when ">="
+        jdsl.method(:eGte)
+      when "<="
+        jdsl.method(:eLte)
+      else
+        raise "Unknown operator #{self.text_value}"
+      end
+    end
+  end
+  
+  module RegExpOperator
+    include Helpers
+    
+    def expr
+      if self.text_value == '!~'
+        jdsl.method(:eRegexNeq)
+      elsif self.text_value == '=~'
+        jdsl.method(:eRegexEq)
+      else
+        raise "Unknown regex operator #{self.text_value}"
+      end
+    end
+  end
+  
+  module BooleanOperator
+    include Helpers
+    
+    def expr
+      case self.text_value
+      when "and"
+        AND_METHOD
+      when "or"
+        OR_METHOD
+      else
+        raise "Unknown operator #{self.text_value}"
+      end
+    end
+  end
+  
+  class Selector < RValue
+    def expr
+      jdsl.eEventValue(source_meta, text_value)
+    end
+  end
+  
+  class SelectorElement < Node;
+    def expr
+      jdsl.eEventValue(source_meta, text_value)
+    end
+  end
+end; end; end; end; end;
```

Comment:
OK for another PR!

Well the previous treetop classes was also complex and hard to read I've hop that splitting them up would help.